ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/task/GetNodeDataFromPeerTask.java

```diff
@@ -83,6 +87,13 @@ protected ResponseStream sendRequest(final EthPeer peer) throws PeerNotConnected
       // Message contains unrequested data, must not be the response to our request.
```

Comment:
The corresponding ticket for this task was badly written and didn't really mention this, but the main focus for this update should be to minimize the number of hash calculations we do to improve efficiency.  

So, I suggest combining this check on line 86 (which checks that the data in the response corresponds to the data requested) with the construction of the map in `mapNodeDataByHash`.  You could do this by writing a for loop that iterates over `nodeData`.  For each element in the list, you can calculate the hash and check if the hash is in `hashes`.  If not, you can return early without calculating all of the hashes.  If it is in the list, you can add it to your map, which will be returned only if you make it all the way through the loop.